File: core/src/drivers/plugins/python/shared/plugin_logger.py
"""
Centralized Plugin Logger Module

Provides a simple, consistent logging interface for OpenPLC Python plugins that
routes log messages through the central logging system:

    C runtime log functions -> Unix socket -> Python log server ->
    REST API -> OpenPLC Editor

This ensures all plugin logs are visible in the Editor's log viewer.

Usage:
    from shared import PluginLogger

    # In plugin init():
    logger = PluginLogger("MODBUS_SLAVE", runtime_args)

    # Throughout plugin:
    logger.info("Server started on port 502")
    logger.error("Connection failed: timeout")
    logger.warn("Retrying connection...")
    logger.debug("Processing request...")

The plugin name is automatically prefixed to all messages, e.g.:
    "[MODBUS_SLAVE] Server started on port 502"
"""

from datetime import datetime, timezone
from typing import Optional
from .safe_logging_access import SafeLoggingAccess


class PluginLogger:
    """
    Thread-safe logger for OpenPLC plugins that routes messages to the
    central logging system.

    Attributes:
        plugin_name: Name of the plugin (used as prefix in log messages)
        is_valid: True if the logger is properly connected to central logging
    """

    def __init__(self, plugin_name: str, runtime_args=None):
        """
        Initialize the plugin logger.

        Args:
            plugin_name: Name of the plugin (e.g., "MODBUS_SLAVE", "MODBUS_MASTER").
                        This will be used as a prefix in all log messages.
            runtime_args: PluginRuntimeArgs instance containing logging function
                         pointers. If None, logger will fall back to print().
        """
        self.plugin_name = plugin_name
        self._prefix = f"[{plugin_name}]"
        self._logging_access: Optional[SafeLoggingAccess] = None
        self._is_valid = False

        if runtime_args is not None:
            self._logging_access = SafeLoggingAccess(runtime_args)
            if self._logging_access.is_valid:
                self._is_valid = True
            else:
                # Log the validation failure via fallback
                self._fallback_print(
                    "WARN",
                    f"SafeLoggingAccess not valid: {self._logging_access.error_msg}. "
                    "Falling back to print()."
                )
                self._logging_access = None

    @property
    def is_valid(self) -> bool:
        """Returns True if logger is connected to central logging system."""
        return self._is_valid

    def _format_message(self, message: str) -> str:
        """Format message with plugin name prefix."""
        return f"{self._prefix} {message}"

    def _fallback_print(self, level: str, message: str):
        """
        Fallback to print when logging access is unavailable.

        This ensures messages are still visible in stdout even if the
        central logging system is not available.
        """
        timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{timestamp}] [{level}] {self._prefix} {message}")

    def info(self, message: str) -> bool:
        """
        Log an informational message.

        Args:
            message: The message to log

        Returns:
            True if message was sent to central logging, False if fallback was used
        """
        formatted = self._format_message(message)
        if self._logging_access:
            success, error_msg = self._logging_access.log_info(formatted)
            if success:
                return True
            # If logging failed, fall back to print
            self._fallback_print("INFO", message)
            return False
        else:
            self._fallback_print("INFO", message)
            return False

    def error(self, message: str) -> bool:
        """
        Log an error message.

        Args:
            message: The message to log

        Returns:
            True if message was sent to central logging, False if fallback was used
        """
        formatted = self._format_message(message)
        if self._logging_access:
            success, error_msg = self._logging_access.log_error(formatted)
            if success:
                return True
            self._fallback_print("ERROR", message)
            return False
        else:
            self._fallback_print("ERROR", message)
            return False

    def warn(self, message: str) -> bool:
        """
        Log a warning message.

        Args:
            message: The message to log

        Returns:
            True if message was sent to central logging, False if fallback was used
        """
        formatted = self._format_message(message)
        if self._logging_access:
            success, error_msg = self._logging_access.log_warn(formatted)
            if success:
                return True
            self._fallback_print("WARN", message)
            return False
        else:
            self._fallback_print("WARN", message)
            return False

    def debug(self, message: str) -> bool:
        """
        Log a debug message.

        Args:
            message: The message to log

        Returns:
            True if message was sent to central logging, False if fallback was used
        """
        formatted = self._format_message(message)
        if self._logging_access:
            success, error_msg = self._logging_access.log_debug(formatted)
            if success:
                return True
            self._fallback_print("DEBUG", message)
            return False
        else:
            self._fallback_print("DEBUG", message)
            return False

File: core/src/drivers/plugins/native/plugin_logger.c
/**
 * @file plugin_logger.c
 * @brief Centralized Plugin Logger Implementation for Native OpenPLC Plugins
 */

#include "plugin_logger.h"
#include "../../plugin_types.h"
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <time.h>

/* Helper macro for formatted stderr output matching human-readable log format */
#define PLUGIN_LOGGER_STDERR(level, fmt, ...)                                                      \
    do                                                                                             \
    {                                                                                              \
        time_t now = time(NULL);                                                                   \
        struct tm t;                                                                               \
        gmtime_r(&now, &t);                                                                        \
        char time_buf[20];                                                                         \
        strftime(time_buf, sizeof(time_buf), "%Y-%m-%d %H:%M:%S", &t);                             \
        fprintf(stderr, "[%s] [%s] " fmt "\n", time_buf, level, ##__VA_ARGS__);                    \
    } while (0)

/* Maximum size for formatted log messages */
#define MAX_LOG_MESSAGE_SIZE 1024

bool plugin_logger_init(plugin_logger_t *logger, const char *plugin_name, void *runtime_args)
{
    if (!logger)
    {
        return false;
    }

    /* Initialize to invalid state */
    logger->is_valid = false;
    logger->log_info = NULL;
    logger->log_debug = NULL;
    logger->log_warn = NULL;
    logger->log_error = NULL;
    logger->plugin_name[0] = '\0';

    if (!plugin_name)
    {
        PLUGIN_LOGGER_STDERR("ERROR", "Plugin logger init failed: plugin_name is NULL");
        return false;
    }

    /* Copy plugin name (with bounds checking) */
    strncpy(logger->plugin_name, plugin_name, sizeof(logger->plugin_name) - 1);
    logger->plugin_name[sizeof(logger->plugin_name) - 1] = '\0';

    if (!runtime_args)
    {
        PLUGIN_LOGGER_STDERR("WARN", "[%s] runtime_args is NULL, logging will fall back to printf",
                             logger->plugin_name);
        return true; /* Still return true - logger will fall back to printf */
    }

    /* Extract logging function pointers from runtime_args */
    plugin_runtime_args_t *args = (plugin_runtime_args_t *)runtime_args;

    logger->log_info = args->log_info;
    logger->log_debug = args->log_debug;
    logger->log_warn = args->log_warn;
    logger->log_error = args->log_error;

    /* Validate that we have at least the basic logging functions */
    if (logger->log_info && logger->log_error)
    {
        logger->is_valid = true;
    }
    else
    {
        PLUGIN_LOGGER_STDERR("WARN", "[%s] Some log functions are NULL, falling back to printf",
                             logger->plugin_name);
    }

    return true;
}

/**
 * @brief Internal helper to format and send log message
 */
static void plugin_logger_log(plugin_logger_t *logger, plugin_log_func_t log_func,
                              const char *level, const char *fmt, va_list args)
{
    char message[MAX_LOG_MESSAGE_SIZE];
    char prefixed_message[MAX_LOG_MESSAGE_SIZE];

    /* Format the user's message */
    vsnprintf(message, sizeof(message), fmt, args);

    /* Add plugin name prefix */
    snprintf(prefixed_message, sizeof(prefixed_message), "[%s] %s", logger->plugin_name, message);

    /* Use central logging if available, otherwise fall back to printf */
    if (log_func)
    {
        log_func("%s", prefixed_message);
    }
    else
    {
        /* Fallback: format to match human-readable log format */
        time_t now = time(NULL);
        struct tm t;
        gmtime_r(&now, &t);
        char time_buf[20];
        strftime(time_buf, sizeof(time_buf), "%Y-%m-%d %H:%M:%S", &t);
        printf("[%s] [%s] [%s] %s\n", time_buf, level, logger->plugin_name, message);
    }
}

void plugin_logger_info(plugin_logger_t *logger, const char *fmt, ...)
{
    if (!logger || !fmt)
    {
        return;
    }

    va_list args;
    va_start(args, fmt);
    plugin_logger_log(logger, logger->log_info, "INFO", fmt, args);
    va_end(args);
}

void plugin_logger_debug(plugin_logger_t *logger, const char *fmt, ...)
{
    if (!logger || !fmt)
    {
        return;
    }

    va_list args;
    va_start(args, fmt);
    plugin_logger_log(logger, logger->log_debug, "DEBUG", fmt, args);
    va_end(args);
}

void plugin_logger_warn(plugin_logger_t *logger, const char *fmt, ...)
{
    if (!logger || !fmt)
    {
        return;
    }

    va_list args;
    va_start(args, fmt);
    plugin_logger_log(logger, logger->log_warn, "WARN", fmt, args);
    va_end(args);
}

void plugin_logger_error(plugin_logger_t *logger, const char *fmt, ...)
{
    if (!logger || !fmt)
    {
        return;
    }

    va_list args;
    va_start(args, fmt);
    plugin_logger_log(logger, logger->log_error, "ERROR", fmt, args);
    va_end(args);
}
